ql/cashflows/iborcoupon.cpp

```diff
@@ -113,56 +118,40 @@ namespace QuantLib {
         */
         Date today = Settings::instance().evaluationDate();
 
-        if (fixingDate_>today)
-            return iborIndex_->forecastFixing(fixingValueDate_,
-                                              fixingEndDate_,
-                                              spanningTime_);
+        if (fixingDate_ > today)
+            return iborIndex_->forecastFixing(fixingValueDate_, fixingEndDate_, spanningTime_);
 
-        if (fixingDate_<today ||
-            Settings::instance().enforcesTodaysHistoricFixings()) {
+        if (fixingDate_ < today || Settings::instance().enforcesTodaysHistoricFixings()) {
             // do not catch exceptions
             Rate result = index_->pastFixing(fixingDate_);
             QL_REQUIRE(result != Null<Real>(),
                        "Missing " << index_->name() << " fixing for " << fixingDate_);
             return result;
         }
 
-        try {
-            Rate result = index_->pastFixing(fixingDate_);
-            if (result!=Null<Real>())
-                return result;
-            else
-                ;   // fall through and forecast
-        } catch (Error&) {
-                ;   // fall through and forecast
-        }
-        return iborIndex_->forecastFixing(fixingValueDate_,
-                                          fixingEndDate_,
-                                          spanningTime_);
+        if (iborIndex_->hasPastFixing(fixingDate_))
+            return iborIndex_->pastFixing(fixingDate_);
```

Comment:
I would remove `hasPastFixing` and leave the try/catch as before.  Calling both `hasPastFixing` and `pastFixing` duplicates most of the work (the check on a valid fixing date, retrieving the time series etc.)
The same goes for `InterestRateIndex`.

Reply:
But you do not mean removing `InterestRateIndex::hasPastFixing(const Date& fixingDate)` itself?

Another approach to avoid code duplicates would be an additional
```cpp
bool InterestRateIndex::hasPastFixing(const Date& fixingDate, Real& pastFixing) const;
```
and than
```cpp
if (iborIndex_->hasPastFixing(fixingDate_, pastFixing))
    return pastFixing;
else
    ...
```

_I really do not like `try {...} catch (...) {...}` to control the program flow. ;-)_

Reply:
I do mean removing `hasPastFixing`.  You can write, for instance,

```
Rate r = iborIndex_->pastFixing();
if (r != Null<Real>())
    return r;
else
    ...
```

You've already done all the work in `pastFixing`, there's no point in doing it twice and make the interface more complex.

Reply:
I don't want to insist on my changes but perhaps some words about my need for it:

I would like to have the `InterestRateIndex::hasPastFixing(fixingDate)` and `FloatingRateCoupon::isFixed()` within the `swig` interface in the end.

In `c#` I need to known if the index and/or coupon is already fixed or if I have to provide the fixing. I am calling `IndexManager::clearHistories();` and `Index::addFixing(...)` quiet often so alternatively I would be happy to have

```cpp
bool IndexManager::hasFixing(const std::string& name, const Date& fixingDate) const;
```

But without the two functions the QuantLib interface might becomes less complex but the calling code becomes far more complex as I have to move the implementation if there is a fixing "far way" from the code which knows best about it.

_And forget about `bool InterestRateIndex::hasPastFixing(const Date& fixingDate, Real& pastFixing) const;`. That's my way of introducing `c#` style like `bool TryGetValue(key, out value)`._